core-bundle/src/Controller/BackendPreviewSwitchController.php

```diff
@@ -90,16 +95,19 @@ public function __construct(FrontendPreviewAuthenticator $previewAuthenticator,
         $this->router = $router;
         $this->tokenManager = $tokenManager;
         $this->csrfTokenName = $csrfTokenName;
+        $this->translator = $translator;
     }
 
     /**
      * @Route("/contao/preview_switch", name="contao_backend_switch")
      */
     public function __invoke(Request $request): Response
     {
-        $user = $this->security->getUser();
+        if (!$this->security->isGranted('ROLE_USER')) {
+            return new Response('Access Denied', $this->security->isGranted('IS_AUTHENTICATED_FULLY') ? Response::HTTP_FORBIDDEN : Response::HTTP_UNAUTHORIZED);
```

Comment:
`$this->denyAccessUnlessGranted('ROLE_USER')` only generates a `403` status. Also I think it's overkill to extend from `AbstractController` just for this. Furthermore I don't think the `PrettyErrorScreenHandler` should be invoked for AJAX requests.

Reply:
yes, it only generates a `403` because apparently Symfony always does that, so I would be fine with not differentiating between `401` and `403`. The PrettyErrorScreenHandler only renders if the request expects an HTML response.

Reply:
Imho if we should leave it as it is.

Reply:
@fritzmg Please don‘t resolve other people‘s conversations unless you actually have resolved their issues (e.g. by merging their suggestion or adjusting the code to their recommendations).

Reply:
> Please don‘t resolve other people‘s conversations unless you actually have resolved their issues (e.g. by merging their suggestion or adjusting the code to their recommendations).

I think we need a better process here then - because then PRs potentially never can get merged.

> I also think that we should use $this->denyAccessUnlessGranted('ROLE_USER') here or at least behave the same (which means not returning a 401 status code).

Please elaborate why, as I still disagree. The response should be 401 if you are not authenticated and 403 if you are authenticated but not allowed to use the preview.

Reply:
Because that is what Symfony‘s `denyAccessUnlessGranted()` method does.

Reply:
But that method is supposed to be only used when you are authenticated, as it will only generate a 403 response. For the same reason we cannot use this method in Contao otherwise, as we want to distinguish between 401 (Not Authenticated) and 403 (Access Denied).

Furthermore, as already explained above, we do not want to throw an exception here as we do not want the prettyerrorscreen listener to show a pretty error screen as this is only used for an AJAX request.

Reply:
As discussed in the Contao call, we want to throw an `AccessDeniedException` here. (Just for the record: `IS_AUTHENTICATED_FULLY` would have had to be replaced with `!IS_AUTHENTICATED`.

Reply:
For the record: I still disagree with this as the exception will now render the full pretty error screen, which I find unnecessary for this AJAX request. But implemented in 77a7ad6fa anyway.

Reply:
Actually, since we are not using `400` for the response to indicate that the username is invalid anymore, I have reverted this to the original state.